File: Coding_practice/BOJ/2618/main.py
def dist1(w1, w2):                          # cur, next
    w1r, w1c = wl1[w1]
    w2r, w2c = wl1[w2]
    return abs(w1r-w2r) + abs(w1c-w2c)


def dist2(w1, w2):
    w1r, w1c = wl2[w1]
    w2r, w2c = wl2[w2]
    return abs(w1r-w2r) + abs(w1c-w2c)


def solve():
    # DP[i][j] -> 1번 경찰차가 마지막으로 i+1번 사건, 2번 경찰차가 마지막으로 j+1 번 사건
    # 이때 마지막 사건을 해결하기 까지 남은 이동거리.
    # row -> 1번, col -> 2번
    # DP[W-1][W-1] = 0
    # DP[W-1][:], DP[:][W-1] 또한 0

    # m, n을 업데이트 할 때 next = max(m, n) + 1
    # dp[m][n] = min(dp[next][n]+dist1, dp[m][next]+dist2)
    # dist1 = dist(m, next)
    # DP initialize
    dp = [[0] * (W+1) for _ in range(W+1)]
    trace = [[0] * (W+1) for _ in range(W+1)]
    for i in range(W):
        dp[W][i] = 0
        dp[i][W] = 0

    # DP update
    for i in range(W-1, -1, -1):             # W-2
        for j in range(i+1):
            n = max(i, j) + 1
            # compare row/col moves
            row_move1 = dp[n][j] + dist1(i, n)      # 1번 자동차가 움직임
            col_move1 = dp[i][n] + dist2(j, n)      # 2번 자동차가 움직임
            row_move2 = dp[n][i] + dist1(j, n)      # 1번 자동차가 움직임
            col_move2 = dp[j][n] + dist2(i, n)      # 2번 자동차가 움직임
            if row_move1 > col_move1:
                trace[i][j] = 2
                dp[i][j] = col_move1
            else:
                trace[i][j] = 1
                dp[i][j] = row_move1

            if row_move2 > col_move2:
                trace[j][i] = 2
                dp[j][i] = col_move2
            else:
                trace[j][i] = 1
                dp[j][i] = row_move2
        # print('#' * N)
        # for row in dp:
        #     print(*row)
        # print('@' * N)
        # for row in trace:
        #     print(*row)
    # for row in dp:
    #     print(*row)
    # print ans
    print(dp[0][0])
    r = c = 0
    for i in range(W):
        print(trace[r][c])
        if trace[r][c] == 2:
            c = max(r, c) + 1
        else:
            r = max(r, c) + 1
    return


N = int(input())
W = int(input())
wl1 = [list(map(int, input().split())) for _ in range(W)]
wl2 = wl1[:]
wl1 = [[1, 1]] + wl1
wl2 = [[N, N]] + wl2
solve()

File: Coding_practice/BOJ/2618/Main.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;

class Main {
    static int BUFFERSIZE = 1000;
    static int[][] buffer = new int[BUFFERSIZE][2];

    private static int dist(int r, int c, int N) {
        // r -> 1st car c -> 2nd car
        if (r == -1) {
            int r1 = 1;
            int c1 = 1;
            int r2 = buffer[c][0];
            int c2 = buffer[c][1];
            return Math.abs(r1-r2) + Math.abs(c1-c2);
        }
        else if (c == -1) {
            int r2 = N;
            int c2 = N;
            int r1 = buffer[r][0];
            int c1 = buffer[r][1];
            return Math.abs(r1-r2) + Math.abs(c1-c2);
        }
        else {
            int r1 = buffer[r][0];
            int c1 = buffer[r][1];
            int r2 = buffer[c][0];
            int c2 = buffer[c][1];
            return Math.abs(r1-r2) + Math.abs(c1-c2);
        }
    }
    private static void print(int[][] arr) {
        for (int[] row : arr) {
            for(int num : row) {
                System.out.printf("%d ", num);
            }
            System.out.println();
        }
    }
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        int W = Integer.parseInt(br.readLine());
        for (int w = 0; w < W; w ++) {
            String[] line = br.readLine().split(" ");
            buffer[w][0] = Integer.parseInt(line[0]);
            buffer[w][1] = Integer.parseInt(line[1]);
        }
        int[][] dp = new int[W+1][W+1];
        int[][] trace = new int[W+1][W+1];
        // update dp. dp[i][j] is... remaining distance.
        // cordinate of the i-th state is buffer[5*i], buffer[5*i+2]
        for (int i = W-1; i >= 0; i--) {
            for (int j = 0; j <= i; j ++) {
                // dp[i][j].
                int n = Math.max(i, j) + 1;
                int cand1 = dp[n][j]+dist(i-1, n-1, N);
                int cand2 = dp[i][n]+dist(n-1, j-1, N);
                int cand3 = dp[n][i]+dist(j-1, n-1, N);
                int cand4 = dp[j][n]+dist(n-1, i-1, N);
                if (cand1 > cand2) {
                    dp[i][j] = cand2;
                    trace[i][j] = 2;
                }
                else {
                    dp[i][j] = cand1;
                    trace[i][j] = 1;
                }
                if (cand3 > cand4) {
                    dp[j][i] = cand4;
                    trace[j][i] = 2;
                }
                else {
                    dp[j][i] = cand3;
                    trace[j][i] = 1;
                }
                // dp[i][j] = Math.min(dp[n][j]+dist(i-1, n-1, N), dp[i][n]+dist(n-1, j-1, N));
                // dp[j][i] = Math.min(dp[n][i]+dist(j-1, n-1, N), dp[j][n]+dist(n-1, i-1, N));
            }
        }
        System.out.println(dp[0][0]);
        int i = 0;
        int j = 0;
        for (int w = 0; w < W; w ++) {
            System.out.println(trace[i][j]);
            int next = Math.max(i, j) + 1;
            if (trace[i][j] == 1) {
                i = next;
            }
            else {
                j = next;
            }
        }
    }
}